File: ising-python-1.0/torus-glauber-heat-bath.py
import random
import math
import time
import sys

def simulation(n, k, b_low, b_high, b_step):
	"""Run the simulation on a Torus of total size n^2. 
	For each value of beta perform k simulations
	"""
	#vary Beta from small to above critical value
	file_name = 'results/torus-heat-bath:' + str(n) + ':k:' + str(k) + ':beta_low:' + str(b_low) + ':beta_high:'
	file_name += str(b_high) + ':beta_step:' + str(b_step) + ':time:' + str(time.time())
	f = open(file_name, 'w')
	beta = b_low
	while beta <= b_high:
		print("Testing Beta = " + str(beta))
		for i in range(0, k):
			print(i)
			iterations, duration = mix_chains(n, beta)
			f.write(str(beta) + ", " + str(iterations) + ", " + str(duration) + "\n")
		beta += b_step
	f.close()

def mix_chains(n, beta):
	"""Run the chains X and Y until they have the same state
	Return the number of moves taken as well as the system time
	"""
	#timing information
	iterations = 0
	start = time.clock()
	#+-1 for spins
	global_diff_count = n * n
	X = [[1 for i in range(n)] for j in range(n)]
	Y = [[-1 for i in range(n)] for j in range(n)]

	shifts = [(1, 0), (-1, 0), (0, 1), (0, -1)] #"shifts" defining the neighbors of a vertex for the torus

	while global_diff_count > 0:
		iterations += 1
		#pick vertex uniformly at random
		v_x, v_y = random.randint(0, n - 1), random.randint(0, n - 1)

		started_same = (X[v_x][v_y] == Y[v_x][v_y])

		#there are 4 neighbors, using torus here
		#determine P(change to +) Y
		Y_pos_prob = 0
		local_spin_sum = 0
		for shift in shifts:
			local_spin_sum += Y[(v_x + shift[0]) % n][(v_y + shift[1]) % n]

		Y_pos_prob = math.exp(beta * local_spin_sum) / (math.exp(beta * local_spin_sum ) + math.exp(-1 * beta * local_spin_sum) )

		#determine P(change to +) for X
		X_pos_prob = 0
		local_spin_sum = 0

		for shift in shifts:
			local_spin_sum += X[(v_x + shift[0]) % n][(v_y + shift[1]) % n]

		X_pos_prob = math.exp(beta * local_spin_sum) / (math.exp(beta * local_spin_sum ) + math.exp(-1 * beta * local_spin_sum) )
		r = random.random()
		
		if r <= Y_pos_prob:
			Y[v_x][v_y] = 1
		else:
			Y[v_x][v_y] = -1
		if r <= X_pos_prob:
			X[v_x][v_y] = 1
		else:
			X[v_x][v_y] = -1

		#update the count diff
		if started_same and X[v_x][v_y] != Y[v_x][v_y]:
			global_diff_count += 1
		elif not started_same and X[v_x][v_y] == Y[v_x][v_y]:
			global_diff_count -= 1

	return (iterations, time.clock() - start)


def main():
	"""
	Main method, read in command line arguments and call simulation
	Command Line Args are:
	size, iterations per beta, lower beta, uper beta, step beta
	"""
	if len(sys.argv) < 6 :
		print("Must supply n, k, b_low, b_high, b_step space delimited")
	else:
		simulation(int(sys.argv[1]), int(sys.argv[2]), float(sys.argv[3]), float(sys.argv[4]), float(sys.argv[5]))

main()

File: ising-c-1.0/torus-glauber-heat-bath.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#define ARC4RANDOM_MAX      0x100000000


void simulation(int n, int k, double b_low, double b_high, double b_step);
int mix_chains(int n, double beta);

/**
 * Main method wrapper
 * @param  argc number of args
 * @param  argv arguments array
 * @return      not used
 */
int main(int argc, char *argv[]){
	if (argc != 6){
		printf("Must supply n, k, b_low, b_high, b_step space delimited");
	}
	int n = atoi(argv[1]);
	int k = atoi(argv[2]);
	double b_low = atof(argv[3]);
	double b_high = atof(argv[4]);
	double b_step = atof(argv[5]);
	simulation(n, k, b_low, b_high, b_step);

}

/**
 * Run the simulation with the specified paramters
 * @param n      The size of the torus
 * @param k      The number of trials to run for each beta
 * @param b_low  The beta to start at
 * @param b_high The beta to end at
 * @param b_step The increment for beta
 */
void simulation(int n, int k, double b_low, double b_high, double b_step){
	double beta = b_low;
	int iterations;
	char file_name[100];
	sprintf(file_name, "results/tours-heat-bath:%d:%d:%f:%f:%f:%d", n, k, b_low, b_high, b_step, (unsigned)time(NULL));
	FILE *f = fopen(file_name, "w");
	if(NULL == f){
		printf("Error opening results file");
		exit(1);
	}
	while(beta <= b_high){
		for(int i = 0; i < k; i++){
			iterations = mix_chains(n, beta);
			fprintf(f, "%f %d\n", beta, iterations);
			printf("beta: %f, k: %d, iterations: %d\n", beta, i, iterations);
		}
		beta += b_step;
	}
	fclose(f);
}

/**
 * Run the chains X and Y until they couple
 * @param  n    The size of the 2D torus
 * @param  beta The value for beta for the partition function
 * @param  X    Pointer for the X chain
 * @param  Y    Pointer for the Y chain
 * @return      The iterations required for coupling
 */
int mix_chains(int n, double beta){
	int X[n][n];
	int Y[n][n];
	unsigned long long iterations = 0;
	/**
	 * use +1 / -1 for spins, stat X at all + and Y at all -
	 */
	
	int global_diff_count = n * n;
	for(int i = 0; i < n; i++){
		for(int j = 0; j < n; j++){
			X[i][j] = 1;
			Y[i][j] = -1;
		}
	}

	int v_x, v_y, started_same, local_spin_sum,
		v_x_new, v_y_new;
	double Y_pos_prob, X_pos_prob, r;

	while(global_diff_count > 0){
		iterations += 1;
		v_x = arc4random_uniform(n);
		v_y = arc4random_uniform(n);

		started_same = X[v_x][v_y] - Y[v_x][v_y];

		local_spin_sum = 0;
		//sum the spins of the neighbors observing wrap around in the Y chain
		v_x_new = (v_x + 1) % n;
		v_y_new = v_y;
		local_spin_sum += Y[v_x_new][v_y_new];
		v_x_new = (v_x - 1) % n;
		if(v_x_new < 0){
			v_x_new += n;
		}
		local_spin_sum += Y[v_x_new][v_y_new];

		v_x_new = v_x;
		v_y_new = (v_y + 1) % n;
		local_spin_sum += Y[v_x_new][v_y_new];
		v_y_new = (v_y - 1) % n;
		if(v_y_new < 0){
			v_y_new += n;
		}
		local_spin_sum += Y[v_x_new][v_y_new];

		//calculate the probability using glauber dynamics for Y chain
		Y_pos_prob = exp(beta * local_spin_sum) / (exp(beta * local_spin_sum ) + exp(-1 * beta * local_spin_sum) );


		local_spin_sum = 0;

		//sum the spins of the neighbors observing wrap around in the X chain
		v_x_new = (v_x + 1) % n;
		v_y_new = v_y;
		local_spin_sum += X[v_x_new][v_y_new];
		v_x_new = (v_x - 1) % n;
		if(v_x_new < 0){
			v_x_new += n;
		}
		local_spin_sum += X[v_x_new][v_y_new];

		v_x_new = v_x;
		v_y_new = (v_y + 1) % n;
		local_spin_sum += X[v_x_new][v_y_new];
		v_y_new = (v_y - 1) % n;
		if(v_y_new < 0){
			v_y_new += n;
		}
		local_spin_sum += X[v_x_new][v_y_new];


		//calculate the probability using glauber dynamics for Y chain
		X_pos_prob = exp(beta * local_spin_sum) / (exp(beta * local_spin_sum ) + exp(-1 * beta * local_spin_sum));


		r = ((double)arc4random() / ARC4RANDOM_MAX);
		
		//update the chains with the proper probabilities 
		if (r <= Y_pos_prob){
			Y[v_x][v_y] = 1;
		}
		else{
			Y[v_x][v_y] = -1;
		}
		if (r <= X_pos_prob){
			X[v_x][v_y] = 1;
		}
		else{
			X[v_x][v_y] = -1;
		}

		//keep track of how many vertexes are different
		if (started_same == 0 && X[v_x][v_y] != Y[v_x][v_y]){
			global_diff_count += 1;
		}
		else if(started_same != 0 && X[v_x][v_y] == Y[v_x][v_y]){
			global_diff_count -= 1;
		}		
	}


	return iterations;

}